client/async.go

```diff
@@ -0,0 +1,266 @@
+	opts ...grpc.DialOption,
+) error {
+	if client != nil {
+		a.client = client
+		return nil
+	}
+	a.logger.WithFields(logrus.Fields{
+		"operation": "configureGRPCForwarderClient",
+	}).Info("connecting to grpc server")
+	dialOpts := append(
+		[]grpc.DialOption{
+			grpc.WithInsecure(),
+			grpc.WithUnaryInterceptor(a.metricsReporterInterceptor),
+		},
+		opts...,
+	)
+	conn, err := grpc.Dial(serverAddress, dialOpts...)
+	if err != nil {
+		return err
+	}
+	a.conn = conn
+	a.client = pb.NewGRPCForwarderClient(conn)
+	return nil
+}
+
+// metricsReporterInterceptor will report metrics from client
+func (a *gRPCClientAsync) metricsReporterInterceptor(
+	ctx context.Context,
+	method string,
+	req interface{},
+	reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	l := a.logger.WithFields(logrus.Fields{
+		"method": method,
+	})
+
+	events := req.(*pb.SendEventsRequest).Events
+
+	defer func(startTime time.Time) {
+		elapsedTime := float64(time.Since(startTime).Nanoseconds() / 1000000)
+		for _, e := range events {
+			metrics.ClientRequestsResponseTime.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+			).Observe(elapsedTime)
+		}
+		l.WithFields(logrus.Fields{
+			"elapsedTime": elapsedTime,
+			"reply":       reply.(*pb.SendEventsResponse),
+		}).Debug("request processed")
+	}(time.Now())
+
+	if err := invoker(ctx, method, req, reply, cc, opts...); err != nil {
+		l.WithError(err).Error("error processing request")
+		for _, e := range events {
+			metrics.ClientRequestsFailureCounter.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+				err.Error(),
+			).Inc()
+		}
+	}
+	for _, e := range events {
+		metrics.ClientRequestsSuccessCounter.WithLabelValues(
+			hostname,
+			method,
+			e.Topic,
+		).Inc()
+	}
+	return nil
+}
+
+func (a *gRPCClientAsync) send(ctx context.Context, event *pb.Event) error {
+	a.wg.Add(1)
+	a.eventsChannel <- event
+	return nil
+}
+
+func (a *gRPCClientAsync) sendRoutine() {
+	ticker := time.NewTicker(a.flushInterval)
+	defer ticker.Stop()
+
+	req := &pb.SendEventsRequest{}
+	req.Events = make([]*pb.Event, 0, a.flushSize)
+
+	send := func() {
+		cpy := req
+		cpy.Id = uuid.NewV4().String()
+		req = &pb.SendEventsRequest{}
+		req.Events = make([]*pb.Event, 0, a.flushSize)
+		go a.sendEvents(cpy, 0)
+	}
+
+	for {
+		select {
+		case e := <-a.eventsChannel:
+			if len(req.Events) == 0 {
+				a.wg.Add(1)
+			}
+			a.wg.Done()
+			req.Events = append(req.Events, e)
+			if len(req.Events) == a.flushSize {
+				send()
+			}
+		case <-ticker.C:
+			if len(req.Events) > 0 {
+				send()
+			}
+		}
+	}
+}
+
+func (a *gRPCClientAsync) sendEvents(req *pb.SendEventsRequest, retryCount int) {
+	l := a.logger.WithFields(logrus.Fields{
+		"operation":  "sendEvents",
+		"requestId":  req.Id,
+		"retryCount": retryCount,
+		"size":       len(req.Events),
+	})
+	if retryCount > a.maxRetries {
+		l.Info("dropped events due to max retries")
```

Comment:
We should report the number of events being dropped.

Reply:
they are reported since "l" has "size"

Reply:
You're right! I didn't see it :)